### PR TITLE
Cart and Product list order now don't change and endpoints renamed

### PR DESCRIPTION
**Pull request by xxk3990**
As someone who has had experience making changes based on user feedback, I noticed the order of the items in the cart and the order of the items in the product catalog would jump around in the order when one is adjusted. In this PR, the order jumping around has been fixed. To test:

PRODUCTS PAGE
1. Add four products to the catalog. Go into the DB and verify that the order of the items in the product catalog corresponds to the created_at dates of each.
2. Modify one and hit submit changes. The product catalog should not change order.
3. Delete one of the products in the middle of the catalog to verify that each item after the one you delete goes up one in the list but the ones before it are unaffected.
4. Add more and try modifying the new ones to verify that although the list can be updated at any point, the date sorting will not be affected and the older items are still appearing before the newer ones.

SHOPPING CART PAGE
1. Add four items. Verify the cart_items list on the FE is sorted by the created_at prop.
2. Increase the quantity and price on some of them and verify that the order of the cart does not change.
3. Delete an item in the middle of the list to verify that the ones after it move up a spot in the list.
4. Add new items to the existing cart to verify that the older items are still appearing before the newer ones.

*Quick note here – in the DB the created_at prop is underscored but on the react FE it's camel cased, which is why I used the camel case version when sorting.

SUCCESS WITH THIS ALSO MEANS THE NEW ENDPOINT NAMES WORK.

### CODE REVIEW

File: client/src/components/Products.js [hunk 1]
```diff
@@ -18,7 +18,21 @@ export default function Products() {
 
     const getProducts = async () => {
         const endpoint = `products`
-        await handleGet(endpoint, setProducts)
+        const host = process.env.REACT_APP_NODE_LOCAL || process.env.REACT_APP_NODE_PROD
+        const url = `${host}/${endpoint}`
```
Comment (paulkogan) [inline client/src/components/Products.js:22]: so the solution was to use these slanted single quotes?

File: client/src/components/Products.js [hunk 2]
```diff
@@ -106,7 +121,8 @@ export default function Products() {
             <h4>No products yet!</h4>
             <section className='add-product'>
                 <span className='product-form-question' id="productname">Product Name: <input type="text" className='user-input' name="productName" value={newProduct.productName} onChange={e => handleChange(e.target.name, e.target.value)} required/></span>
-                <span className='product-form-question' id="productimageUrl">Product Image: <input type="text" className='user-input' id="product-img" name="imageURL" value={newProduct.imageURL} onChange={e => handleChange(e.target.name, e.target.value)} required placeholder='Please paste a proper link.'/></span>
+                <span className='product-form-question' id="product-image-url">Product Image: <input type="text" className='user-input' id="product-img" name="imageURL" value={newProduct.imageURL} onChange={e => handleChange(e.target.name, e.target.value)} required placeholder='Please paste a proper link.'/></span>
```
Comment (paulkogan) [inline client/src/components/Products.js]: can you format blocks like this so each attribute in a JSX element is on a separate line?
type="text" 
className='user-input' 
id="product-img" 
name="imageURL" 
value={newProduct.imageURL} 
onChange={e => handleChange(e.target.name, e.target.value)}

File: client/src/components/ShoppingCart.js [hunk 3]
```diff
@@ -31,7 +31,9 @@ export default function ShoppingCart() {
                 setCartItems([]); 
             } else {
                 setCartTotal(responseData.cart_total)
-                setCartItems(responseData.cart_items) //set it equal to data from API
+                //sort response data by place in cart (newer items at the end) so order can't change
```
Comment (paulkogan) [inline client/src/components/ShoppingCart.js]: Oh, I see the place in cart idea. 
The idea of sorting is the right solution, so kudos on that. 

But try to  avoid adding a new field just for sorting.
You always want to keep things as simple as possible.
What about sorting by another field?

File: client/src/components/ShoppingCart.js [hunk 4]
```diff
@@ -47,7 +49,7 @@ export default function ShoppingCart() {
 
 
     const addCartItem = async(item) => {
-        const endpoint = `addToCart`
+        const endpoint = `cartItems`
```
Comment (paulkogan) [inline client/src/components/ShoppingCart.js]: Dont use camelCase in naming endpoints. Use lowercase as oneword or spinal-case.
so here: cart-items or cartitems

https://blog.restcase.com/5-basic-rest-api-design-guidelines/

File: client/src/components/ShoppingCart.js [hunk 5]
```diff
@@ -160,7 +161,7 @@ export default function ShoppingCart() {
     }
 
     const submitOrder = async() => {
-        const endpoint = `submitOrder`;
+        const endpoint = `orders`;
```
Comment (paulkogan) [inline client/src/components/ShoppingCart.js:163]: this is good

File: server/controllers/cartItems-controller.js [hunk 6]
```diff
@@ -35,6 +35,7 @@ const createCartItem = async (req, res) => {
             product_uuid: req.body.product_uuid,
             price: req.body.price,
             image_url: req.body.image_url,
+            place_in_cart: req.body.place_in_cart,
```
Comment (paulkogan) [inline server/controllers/cartItems-controller.js]: See above comment about place_in_cart
I see you did a lot of work on it BUT... its not a great pattern.
Can you think of another field to sort by?
Without adding an extra field. Keep things always as simple as needed.

File: server/controllers/cartItems-controller.js [hunk 7]
```diff
@@ -48,7 +49,7 @@ const createCartItem = async (req, res) => {
 const updateCartItem = async (req, res, next) => {
     const itemToUpdate = await models.Cart_Item.findOne({
         where: {
-            "product_uuid": req.body.item.product_uuid
+            "uuid": req.body.item.uuid
```
Comment (paulkogan) [inline server/controllers/cartItems-controller.js:47]: So is this a switch in naming or in usage?
The payload is send a product_uuid or an item_uuid?

ideally, all you'd need is the item_uuid.

File: client/src/components/Products.js (hunk 1 above)
Comment (xxk3990) [reply client/src/components/Products.js:22]: Do you mean for the connections issue we had? I think the issue we had during our call was that we forgot to restart the server when we changed the .env prop values. I had completely forgotten that React requires a server restart if env variables are changed, which might be why it didn't work for you. My mistake. The solution for the list order changing (on both the products page and the shopping cart page) was to sort the array using a place_in_catalog (products) or place_in_cart (cart_items).

File: server/migrations/products-migration.js [hunk 8]
```diff
@@ -12,6 +12,7 @@ module.exports = {
       product_name: Sequelize.STRING,
       image_url: Sequelize.STRING,
       price: Sequelize.INTEGER,
+      place_in_catalog: Sequelize.INTEGER,
```
Comment (paulkogan) [inline server/migrations/products-migration.js]: same with place in catalog? Can you find an already existing field you can sort by?

File: server/router.js [hunk 9]
```diff
@@ -8,17 +8,17 @@ const router = (app) => {
...
     app.get('/orders', order.userOrders) //this one is for specific users, need another for admin purposes
     //app.get('/allOrders', admin.getAllOrders) not created yet
-    app.post('/submitOrder', order.submitOrder)
+    app.post('/orders', order.submitOrder)
     app.get('/orderItems', items.getOrderItems);
     app.get('/cartItems', cart.getCartItems)
-    app.post("/addToCart", cart.createCartItem);
-    app.put('/updateCartItem', cart.updateCartItem);
-    app.delete("/deleteCartItem", cart.deleteCartItem)
+    app.post("/cartItems", cart.createCartItem);
```
Comment (paulkogan) [inline server/router.js]: OK, its good to see all teh endpoints in one shot.

You just need to change cartItems => cart-items or. cartitems

**Review by paulkogan — CHANGES_REQUESTED** [review]
Ok, left you some notes

File: client/src/components/ShoppingCart.js (hunk 3 above)
Comment (xxk3990) [reply client/src/components/ShoppingCart.js]: I wonder if I can sort by created_at. Will try and if I'm successful I'll merge it into develop and then update the review here

Comment (xxk3990) [reply client/src/components/ShoppingCart.js]: I was able to do it using the createdAt field! Committed my changes to develop and edited the steps to recreate to reflect the change

File: server/controllers/cartItems-controller.js (hunk 7 above)
Comment (xxk3990) [reply server/controllers/cartItems-controller.js:47]: It used to be the product uuid but then I changed it to the cart_item uuid to be safer.

File: server/controllers/products-controller.js [hunk 10]
```diff
@@ -14,13 +14,14 @@ const getProducts = async (req, res) => {
...
-    const priceAsInt = utils.removeDecimalIfNeeded(req.body.price) //remove decimal entered on FE
+    const priceAsInt = utils.removeDecimalOrAddZeros(req.body.price) //remove decimal entered on FE
     const newProduct = {
         //id, product_name, image_url, price
         uuid: uuidv4(),
         product_name: req.body.product_name,
         image_url: req.body.image_url, //replace with AWS link later on
-        price: priceAsInt
+        price: priceAsInt,
+        place_in_catalog: req.body.place_in_catalog
```
Comment (paulkogan) [inline server/controllers/products-controller.js]: so what is "place_in_catalog"?

File: client/src/components/Products.js (hunk 1 above)
Comment (paulkogan) [reply client/src/components/Products.js:22]: So why use `products` instead of 'products'

**Review by paulkogan — COMMENTED** [review]
just last two questions...

File: server/controllers/products-controller.js (hunk 10 above)
Comment (xxk3990) [reply server/controllers/products-controller.js]: That was my fault, forgot to delete it here. Had that change in but hadn't committed it yet. Just added

**Pull request merged**